Dockerfile

```diff
@@ -24,7 +24,7 @@ RUN useradd ${PG_USER} -d ${PG_HOME} && \
 #-------------------------------------------------------------------------------
 WORKDIR ${PG_HOME}
 
-COPY --chown=postgres:postgres lib/ ./lib/
+COPY --chown=postgres:postgres lib/postgres-xl ./lib/postgres-xl
```

Comment:
Why? I see e3dc6d6c7ab6bf95f29251eab343d57d43bb707a says it might be used for something else—except it isn't:
https://github.com/tiredpixel/postgres-xl-docker/tree/master/lib
And even if it were, would we not want all the library dependencies present in the image? How else would it be compiled?

Reply:
1. I've tried to add `pgpool` as submodule. It works as separate image and it doesn't need to add to the image.

2. To be honest I hadn't seen that target directory has path `/var/lib/postgresql/lib/`. I had tried to make `pgpool` as part of the image early. But it's strange that path for `pgpool` or `wait-for-it` will consist `/var/lib/postgresql/lib/` (because it is not postgresql). Maybe we need to change it.

    For ex `postgres-xl` will be copied to `/var/lib/postgres-xl`, `wait` – `/var/lib/wait`

    It isn't need to decide instead of user. Let he own decide what he want to copy to the image and where it needs. What do think about that?

3. Why we need user in the image? Docker has the idea **single service as a container**. Or couldn't I understand your implementation?

Reply:
1-2 yes, **I agree with you. Pg-pool is optional.** We can just write about it in the documentation (like tutorial).

What about 3 I don't still know for what we use **users politics in Docker file**.

And last I am convinced that we need to copy lib dependencies separately to the image (without any sub directory). I suggest copy `/lib/postgres-xl` to `/var/lib/postgres-xl`.

Reply:
If we agree that Pgpool is optional, we can certainly mention it in the README—maybe along with HAProxy, etc. But in that case, there's no need for us to include it in the example configuration, and certainly not to include the 750-line example configuration file.

Sorry, I don't understand your question about 3. What user's politics? Is it possible that's a typo? I'm not aware of any politics in the Dockerfile. The user itself is because of least privilege, so it is compiled and ultimately runs as `postgres` user—similar to PostgreSQL itself (i.e. non-Postgres-XL).

Are you aware that there are *two* Docker images, here? The `Dockerfile` is used to install all the build dependencies for development, as well as to include the Postgres-XL source (it has to come from somewhere, after all). But those are not included in the final images released to Docker Hub; those are built by `ci/Dockerfile`, are are not expected to be easily runnable locally by a user (since they depend on a build system I host privately). So, it's not that the `lib/postgres-xl` files go into the final image (if you download the Docker Hub pacakged images, you'll see this); the source files are compiled, and then those compiled binaries are included in the other Docker image. Does that answer your question?

Reply:
No, I mean main Dockerfile. I we have politics that one process in one container why we use OS users? Why we use users inside docker?

```Dockerfile
RUN useradd ${PG_USER} -d ${PG_HOME} && \
    mkdir -p ${PG_LIB} ${PG_HOME} && \
    chown -R ${PG_USER}:${PG_USER} ${PG_LIB} ${PG_HOME}

...

USER ${PG_USER}

...

USER root

...

USER ${PG_USER}

```

What about CI I need more time to understand it.

Reply:
Because Docker has users within the container… and dropping root privileges is standard practice from a security point of view. And this isn't an OS user; it's a user within the image.

It sounds like you might like to spend some time looking at things like the Postgres images, if you haven't already:
https://github.com/docker-library/postgres/blob/master/12/Dockerfile